nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -158,75 +160,81 @@ static int nntrainer_model_push_data(const GstTensorTrainerFramework *fw,
   ml_logd("<called>");
   ml_logd("pid[%d], tid[%d]", pid, tid);
 
-  if (!nntrainer) {
-    ml_loge("Failed get nntrainer");
+  if (!trainer) {
+    ml_loge("Failed get trainer");
     return -1;
   }
 
-  nntrainer->num_push_data++;
-  ml_logd("Received data (%d/%d(total))", nntrainer->num_push_data,
-          nntrainer->total_num_samples);
-  if (nntrainer->total_num_samples < nntrainer->num_push_data) {
+  trainer->incrementNumOfPushedSamples();
+  ml_logd("Received data (%d/%d(total))", trainer->getNumOfPushedSamples(),
+          trainer->getNumOfTotalSamples());
+  if (trainer->getNumOfTotalSamples() < trainer->getNumOfPushedSamples()) {
     ml_logd("Already received all data required for the train, "
             "but push_data is called");
     return 0;
   }
 
-  if (nntrainer->train_data->push_count < nntrainer->num_training_samples) {
-    data = nntrainer->train_data.get();
+  if (trainer->train_tensors_queue->getPushCount() <
+      trainer->getNumOfTrainingSamples()) {
+    tensors_queue = trainer->train_tensors_queue.get();
     ml_logd("#### T-Data ####");
-  } else if (nntrainer->valid_data->push_count <
-             nntrainer->num_validation_samples) {
-    data = nntrainer->valid_data.get();
+  } else if (trainer->valid_tensors_queue->getPushCount() <
+             trainer->getNumOfValidationSamples()) {
+    tensors_queue = trainer->valid_tensors_queue.get();
     ml_logd("#### V-Data ####");
   } else {
     ml_loge("Invalid push_count");
     return -1;
   }
 
-  ml_logd("number of inputs(%d) and labels(%d)", nntrainer->num_inputs,
-          nntrainer->num_labels);
+  tensors_queue->push(input);
 
-  std::unique_lock<std::mutex> lock(data->queue_lock);
-  data->data_full.wait(lock, [&] { return !data->isQueueFull(); });
+  ml_logd("T-pushed: %d/%d, V-pushed:%d/%d\n",
+          trainer->train_tensors_queue->getPushCount(),
+          trainer->getNumOfTrainingSamples(),
+          trainer->valid_tensors_queue->getPushCount(),
+          trainer->getNumOfValidationSamples());
+  ml_logd("<leaved>");
+  return 0;
+}
+
+void SubPlugin::TensorsQueue::push(const GstTensorMemory *input) {
+  ml_logd("<called>");
+
+  std::unique_lock<std::mutex> lock(queue_lock);
```

Comment:
Q1: Does ```unique_lock``` destroy when the control goes out of scope?
It is unlocked at the line 231.

Q2: Which part do you try to guard?

Reply:
Q1: yes
Q2: nnstreamer writes data to a ```std::Vector<TensorsData> queue``` and nntrainer reads data from the queue.